performances/src/profiling.ts

```diff
@@ -0,0 +1,159 @@
+import { CDPSession, Page, Protocol } from 'puppeteer'
+import { ProfilingResults, ProfilingOptions } from './types'
+
+export async function startProfiling(options: ProfilingOptions, page: Page) {
+  const client = await page.target().createCDPSession()
+  const stopCPUProfiling = await startCPUProfiling(options, client)
+  const { stopMemoryProfiling, takeMemoryMeasurements } = await startMemoryProfiling(options, client)
+  const stopNetworkProfiling = await startNetworkProfiling(options, client)
+
+  return {
+    takeMeasurements: async () => {
+      await takeMemoryMeasurements()
+    },
+    stopProfiling: async (): Promise<ProfilingResults> => ({
+      memory: await stopMemoryProfiling(),
+      cpu: await stopCPUProfiling(),
+      ...stopNetworkProfiling(),
+    }),
+  }
+}
+
+async function startCPUProfiling(options: ProfilingOptions, client: CDPSession) {
+  await client.send('Profiler.enable')
+  await client.send('Profiler.start')
+
+  return async () => {
+    const { profile } = await client.send('Profiler.stop')
+
+    const timeDeltaForNodeId = new Map<number, number>()
+
+    for (let index = 0; index < profile.samples!.length; index += 1) {
+      const nodeId = profile.samples![index]
+      timeDeltaForNodeId.set(nodeId, (timeDeltaForNodeId.get(nodeId) || 0) + profile.timeDeltas![index])
+    }
+
+    let totalConsumption = 0
+    let sdkConsumption = 0
+    for (const node of profile.nodes) {
+      const consumption = timeDeltaForNodeId.get(node.id) || 0
+      totalConsumption += consumption
+      if (isSdkUrl(options, node.callFrame.url)) {
+        sdkConsumption += consumption
+      }
+    }
+
+    return { total: totalConsumption, sdk: sdkConsumption }
+  }
+}
+
+async function startMemoryProfiling(options: ProfilingOptions, client: CDPSession) {
+  await client.send('HeapProfiler.enable')
+  await client.send('HeapProfiler.startSampling', {
+    // Set a low sampling interval to have more precise measurement
+    samplingInterval: 100,
+  })
+
+  const measurements: Array<{ sdkConsumption: number; totalConsumption: number }> = []
+
+  return {
+    takeMemoryMeasurements: async () => {
+      await client.send('HeapProfiler.collectGarbage')
+      const { profile } = await client.send('HeapProfiler.getSamplingProfile')
+
+      const sizeForNodeId = new Map<number, number>()
```

Comment:
I wonder why using a `Map` here instead of a simple object `{}`, is there a benefit?

Reply:
I'd like to ask, is there any benefit to use `{}` instead of `Map` ? 😃 

I tend to use `Map` wherever possible, because using `{}` has too many subtleties to worry about, like:
* if my key is not a string, is it safe to cast it as a string?
* what if my key clashes with object methods? should I use `map.hasOwnProperty(key)`, `key in map`, `Object.prototype.hasOwnProperty.call(map, key)`?
* what happen if the iteration order changes?

Even if those questions may not apply in this particular use-case, I find that it's good practice to always use `Map` instead of `{}` when indexing values in a map.

Reply:
Interesting, thanks for the explanation, I was aware of the `hasOwnProperty` thing, but not others.